Userland/Shell/PosixParser.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2022, Ali Mohammad Pur <mpfard@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <Shell/AST.h>
+#include <Shell/PosixLexer.h>
+
+namespace Shell::Posix {
+
+class Parser {
+public:
+    Parser(StringView input, bool interactive = false)
+        : m_lexer(input)
+        , m_in_interactive_mode(interactive)
+        , m_eof_token(Token::eof())
+    {
+        fill_token_buffer();
+    }
+
+    RefPtr<AST::Node> parse();
+    RefPtr<AST::Node> parse_word_list();
+
+    struct Error {
+        DeprecatedString message;
+        Optional<AST::Position> position;
+    };
+    auto& errors() const { return m_errors; }
+
+private:
+    Optional<Token> next_expanded_token();
+    Vector<Token> perform_expansions(Vector<Token> tokens);
+    void fill_token_buffer();
+
+    Token const& peek() const
+    {
+        if (eof())
+            return m_eof_token;
+        return m_token_buffer[m_token_index];
+    }
+    Token const& consume()
+    {
+        if (eof())
```

Comment:
not sure if the following logic would not help us catch more errors:
* always allow bumping the index,
* return eof as currently implemented,
* throw an error if index is beyond buffer size…

similarly `peek` and `skip`—thatʼs just an idea, so ignore if not state of the art for parsing… 😬

Reply:
In general, it's always nicer to handle eof in the specific parse rule - otherwise you'll end up with the most generic error in the world, which are almost always useless.

e.g. `Expected 'Esac', not eof` instead of `Unexpected eof`